controllers/provider-alicloud/pkg/controller/worker/machines.go

```diff
@@ -40,13 +44,94 @@ func (w *workerDelegate) MachineClassList() runtime.Object {
 	return &machinev1alpha1.AlicloudMachineClassList{}
 }
 
+// getSeedCloudProviderCredentials gets Seed's cloud provider's credentials from BackupBucket SecretRef
+func (w *workerDelegate) getSeedCloudProviderCredentials(ctx context.Context) (*alicloud.Credentials, error) {
+	backupBuckets := &extensionsv1alpha1.BackupBucketList{}
+	err := w.client.List(ctx, backupBuckets)
+	if err != nil {
+		return nil, err
+	}
+	credentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, w.client, &backupBuckets.Items[0].Spec.SecretRef)
```

Comment:
Well, let's not introduce such an implicit, error-prone contract. What if there is no backup configured for this seed?
Instead, if you need the Alicloud credentials, then let the operator add them explicitly to your Alicloud extension controller.

Reply:
@rfranzke, so let's add secrets in https://github.wdf.sap.corp/kubernetes/landscape-setup/tree/master/components/gardener-extensions/provider-alicloud/resources.

Reply:
Why do you post internal links to the public repository? 

> I also tried to add a Secret to https://github.com/gardener/gardener-extensions/tree/master/controllers/provider-alicloud/charts/provider-alicloud/templates, but there is no way for controller registration to config credention information. Credential information should always be stored in Secret.

Yes, but we don't have such mechanism yet, and we probably also don't want to invent one. The idea is to put the credentials as part of the Helm chart values in the `ControllerRegistration`, and advise landscape operators to not only encrypt `Secret`s but also these resources.

Reply:
I wiped off the internal link.